sys/mutex.c

```diff
@@ -0,0 +1,44 @@
+#include <sync.h>
+#include <atomic.h>
+#include <mutex.h>
+
+#define MTX_UNOWNED 0
+
+static bool mtx_owned(mtx_t *mtx) {
+  return mtx->mtx_state != MTX_UNOWNED;
+}
+
+static thread_t *mtx_owner(mtx_t *mtx) {
+  return (thread_t *)mtx->mtx_state;
+}
+
+static int mtx_try_lock(mtx_t *mtx) {
+  return atomic_cmp_exchange(&mtx->mtx_state, MTX_UNOWNED,
+                             (uint32_t)thread_self());
+}
+
+void mtx_init(mtx_t *mtx) {
+  mtx->mtx_state = MTX_UNOWNED;
+  turnstile_init(&mtx->turnstile);
+}
+
+void mtx_lock(mtx_t *mtx) {
+  /* TODO: Implement recursive mutexes */
+  assert(mtx_owner(mtx) != thread_self());
+
+  while (!mtx_try_lock(mtx)) {
+    cs_enter();
+    /* Check if the mutex got unlocked since a call to mtx_try_lock */
+    if (mtx->mtx_state == MTX_UNOWNED) {
+      cs_leave();
+      continue;
+    }
+    assert(mtx_owned(mtx));
+    turnstile_wait(&mtx->turnstile);
```

Comment:
I have carefully examined that, and I think it does not. After `cs_leave` and before `sched_yield` if there is unvoluntary sched_yield() the thread won't get added to run queue, but will wait to get signaled. The funny part is that it will yield again after that and that is actually pretty pointless yield, I think we need scheduler function which yields AND leaves critical section.